packages/chat-stateful-client/src/ResourceDownloadQueue.ts

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+/* @conditional-compile-remove(teams-inline-images-and-file-sharing) */
+import { ChatContext } from './ChatContext';
+/* @conditional-compile-remove(teams-inline-images-and-file-sharing) */
+import { ChatError } from './ChatClientState';
+/* @conditional-compile-remove(teams-inline-images-and-file-sharing) */
+import { ChatMessageWithStatus } from './types/ChatMessageWithStatus';
+/* @conditional-compile-remove(teams-inline-images-and-file-sharing) */
+import { CommunicationTokenCredential } from '@azure/communication-common';
+
+/* @conditional-compile-remove(teams-inline-images-and-file-sharing) */
+/**
+ * @private
+ */
+export class ResourceDownloadQueue {
+  private _messageQueue: ChatMessageWithStatus[] = [];
+  private _context: ChatContext;
+  private _isRequesting = false;
+  private _credential: CommunicationTokenCredential;
+
+  constructor(context: ChatContext, credential: CommunicationTokenCredential) {
+    this._context = context;
+    this._credential = credential;
+  }
+
+  public isRequesting(): boolean {
+    return this._isRequesting;
+  }
+
+  public containsMessage(message: ChatMessageWithStatus): boolean {
+    let contains = false;
+    if (this._messageQueue.find((m) => m.id === message.id)) {
+      contains = true;
+    }
+    return contains;
+  }
+
+  public addMessage(message: ChatMessageWithStatus): void {
+    // make a copy of message and add to queue
+    const copy = { ...message };
```

Comment:
Add message could encapsulate duplicate checking, rather than needing to be done outside the class.

Reply:
I don't agree with this being the responsibility of the queue. If the caller for whatever reason wants to put in a duplicate entry the queue shouldn't be preventing that.